packages/workers/src/block-indexer.ts

```diff
@@ -0,0 +1,173 @@
+import { ChainName, type Chain } from '@metamask-institutional/defi-adapters'
+import { ChainIdToChainNameMap } from '@metamask-institutional/defi-adapters/dist/core/constants/chains.js'
+import type { CustomJsonRpcProvider } from '@metamask-institutional/defi-adapters/dist/core/provider/CustomJsonRpcProvider.js'
+
+import { AVERAGE_BLOCKS_PER_10_MINUTES } from '@metamask-institutional/defi-adapters/dist/core/constants/AVERAGE_BLOCKS_PER_10_MINS.js'
+import { AVERAGE_BLOCKS_PER_DAY } from '@metamask-institutional/defi-adapters/dist/core/constants/AVERAGE_BLOCKS_PER_DAY.js'
+import { logger } from './logger.js'
+
+export class BlockRunner {
+  private _provider: CustomJsonRpcProvider
+  private _chainId: Chain
+  private _chainName: string
+
+  private static readonly _BATCH_SIZE = 50
+  private static readonly _MAX_RETRIES = 5
+
+  private _latestBlockNumber: number | undefined
+  private _getStartBlockNumber: () => Promise<number>
+  private _processBlockFn: (blockNumber: number) => Promise<void>
+  private _onError: (latestSafeProcessedBlock: number) => Promise<void>
+
+  constructor({
+    provider,
+    chainId,
+    getStartBlockNumberFn,
+    processBlockFn,
+    onError,
+  }: {
+    provider: CustomJsonRpcProvider
+    chainId: Chain
+    getStartBlockNumberFn: () => Promise<number>
+    processBlockFn: (blockNumber: number) => Promise<void>
+    onError: (latestSafeProcessedBlock: number) => Promise<void>
+  }) {
+    this._chainName = ChainName[chainId]
+    this._processBlockFn = processBlockFn
+    this._getStartBlockNumber = getStartBlockNumberFn
+    this._onError = onError
+
+    this._provider = provider
+    this._chainId = chainId
+    this._chainName = ChainIdToChainNameMap[chainId]
+  }
+
+  private async waitForNewBlockToProcess(currentBlock: number): Promise<void> {
+    let backoff = 1000 // Start with 1 second
+
+    if (!this._latestBlockNumber) {
+      this._latestBlockNumber = await this._provider.getBlockNumber()
+    }
+
+    while (currentBlock >= this._latestBlockNumber) {
+      try {
+        const blockNumber = await this._provider.getBlockNumber()
+        logger.info(`[${this._chainName}] Latest block number: ${blockNumber}`)
+        if (blockNumber > this._latestBlockNumber) {
+          this._latestBlockNumber = blockNumber
+          return
+        }
+      } catch (error) {
+        logger.error(`[${this._chainName}] Error fetching block number:`, error)
+      }
+      await new Promise((resolve) => setTimeout(resolve, backoff))
+      backoff = Math.min(backoff * 2, 60000) // Exponential backoff up to max 1 minute
+    }
+  }
+
+  async start() {
+    logger.info(`[${this._chainName}] Starting block indexer...`)
+    let processingBlockNumber = await this._getStartBlockNumber()
+    this._latestBlockNumber = await this._provider.getBlockNumber()
+    let retryCount = 0
+
+    while (retryCount <= BlockRunner._MAX_RETRIES) {
```

Comment:
Why are they a maximum of retries?

We do not want the process to ever stop, even if it keeps failing. We can then rely on errors logs to alert us know if the process hasn't been healthy for a while.